api/rest/client/client.go

```diff
@@ -70,13 +79,32 @@ func NewClient(cfg *Config) (*Client, error) {
 		urlPrefix += "http://"
 	}
 
-	if cfg.APIAddr == nil {
+	if cfg.Timeout == 0 {
+		cfg.Timeout = DefaultTimeout
+	}
+
+	// When no host/port/multiaddress defined, we set the default
+	if cfg.APIAddr == nil && cfg.Host == "" && cfg.Port == "" {
 		cfg.APIAddr, _ = ma.NewMultiaddr(DefaultAPIAddr)
 	}
-	_, host, err := manet.DialArgs(cfg.APIAddr)
-	if err != nil {
-		return nil, err
+
+	var host string
+	// APIAddr takes preference. If it exists, it's resolved and dial args
+	// extracted. Otherwise, host port is used.
+	if cfg.APIAddr != nil {
+		// Resolve multiaddress just in case and extract host:port
+		resolveCtx, cancel := context.WithTimeout(ctx, cfg.Timeout)
+		defer cancel()
+		resolved, err := madns.Resolve(resolveCtx, cfg.APIAddr)
+		cfg.APIAddr = resolved[0]
```

Comment:
Just double checking that the first element of resolved is what we want here.  Looking at madns I found that the first element in the case of an ipv4 address is the ip4maddr from: `ip4maddr, err := ma.NewMultiaddr("/ip4/" + ip4.String())`.  I had trouble tracking down what `ip4` exactly is from the code but was suspicious this might just be an ip and not include a transport.  Maybe that is fine too?  Just bringing up because that wouldn't match with the default multiaddr (ip + tcp addrs).

Reply:
I guess you saw:

```
		ip4maddr, err := ma.NewMultiaddr("/ip4/" + ip4.String())
		if err != nil {
			return result, err
		}
               parts := append([]ma.Multiaddr{ip4maddr}, encap...)
               result = append(result, ma.Join(parts...))
```

See it does ma.Join on the result, so the resulting multiaddress has all the parts after the "/ip4/ip", including the /tcp/port etc..

Reply:
Ah I was thrown off thinking result result already has something else in it, thanks this makes sense